src/xrpld/app/tx/detail/MPTokenAuthorize.cpp

```diff
@@ -0,0 +1,262 @@
+        if (ctx.tx.getFlags() & tfMPTUnauthorize)
+        {
+            if (!sleMpt)
+                return tecOBJECT_NOT_FOUND;
+
+            if ((*sleMpt)[sfMPTAmount] != 0)
+            {
+                auto const sleMptIssuance = ctx.view.read(
+                    keylet::mptIssuance(ctx.tx[sfMPTokenIssuanceID]));
+                assert(sleMptIssuance);
+                if (!sleMptIssuance)
+                    return tefINTERNAL;
+
+                return tecHAS_OBLIGATIONS;
+            }
+
+            return tesSUCCESS;
+        }
+
+        // Now test when the holder wants to hold/create/authorize a new MPT
+        auto const sleMptIssuance =
+            ctx.view.read(keylet::mptIssuance(ctx.tx[sfMPTokenIssuanceID]));
+
+        if (!sleMptIssuance)
+            return tecOBJECT_NOT_FOUND;
+
+        if (accountID == (*sleMptIssuance)[sfIssuer])
+            return tecNO_PERMISSION;
+
+        // if holder wants to use and create a mpt
+        if (sleMpt)
+            return tecDUPLICATE;
+
+        return tesSUCCESS;
+    }
+
+    auto const sleMptIssuance =
+        ctx.view.read(keylet::mptIssuance(ctx.tx[sfMPTokenIssuanceID]));
+    if (!sleMptIssuance)
+        return tecOBJECT_NOT_FOUND;
+
+    std::uint32_t const mptIssuanceFlags = sleMptIssuance->getFieldU32(sfFlags);
+
+    // If tx is submitted by issuer, they would either try to do the following
+    // for allowlisting:
+    // 1. authorize an account
+    // 2. unauthorize an account
+    //
+    // Note: `accountID` is issuer's account
+    //       `holderID` is holder's account
+    if (accountID != (*sleMptIssuance)[sfIssuer])
+        return tecNO_PERMISSION;
+
+    // If tx is submitted by issuer, it only applies for MPT with
+    // lsfMPTRequireAuth set
+    if (!(mptIssuanceFlags & lsfMPTRequireAuth))
+        return tecNO_AUTH;
+
+    // The holder must create the MPT before the issuer can authorize it.
+    if (!ctx.view.exists(
+            keylet::mptoken(ctx.tx[sfMPTokenIssuanceID], *holderID)))
+        return tecOBJECT_NOT_FOUND;
+
+    return tesSUCCESS;
+}
+
+TER
+MPTokenAuthorize::authorize(
+    ApplyView& view,
+    beast::Journal journal,
+    MPTAuthorizeArgs const& args)
+{
+    auto const sleAcct = view.peek(keylet::account(args.account));
+    if (!sleAcct)
+        return tecINTERNAL;
+
+    // If the account that submitted the tx is a holder
+    // Note: `account_` is holder's account
+    //       `holderID` is NOT used
+    if (!args.holderID)
+    {
+        // When a holder wants to unauthorize/delete a MPT, the ledger must
+        //      - delete mptokenKey from owner directory
+        //      - delete the MPToken
+        if (args.flags & tfMPTUnauthorize)
+        {
+            auto const mptokenKey =
+                keylet::mptoken(args.mptIssuanceID, args.account);
+            auto const sleMpt = view.peek(mptokenKey);
+            if (!sleMpt || (*sleMpt)[sfMPTAmount] != 0)
+                return tecINTERNAL;
```

Comment:
The rule that I observ is that errors in `apply` are generally tecINTERNAL since the these errors should've already been caught in the preclaim in the first place, so we'd never expect this condition be triggered.

Reply:
I don't think that's a good reason.

Reply:
I think this is practice the codebase currently has. Perhaps @ximinez can weigh in

Reply:
IMHO, if something has already been checked in an earlier step (like `preflight` or `preclaim`), and thus should be impossible, then our practice has been to return `tef/tecINTERNAL` in later steps. 

Think of it more like an `assert`: We want to know something went wrong, but we don't want to crash the server. Particularly with `tecINTERNAL`, if we see that on a validated ledger, we know something bad happened, and we can replay the transaction to see what happened and fix it. If we just return the same error codes that `preclaim()` returned, we'd never know there was a problem.